File: app/src/main/scala/tripaint/grid/ImageChange.scala
package tripaint.grid

import tripaint.Color
import tripaint.coords.TriangleCoords
import tripaint.image.ImageStorage
import tripaint.model.Change

import scala.collection.mutable.ArrayBuffer

class ImageChange private (image: ImageStorage, pixelsChanged: Seq[ImageChange.PixelChange])
    extends Change {
  def redo(): Unit =
    for ch <- pixelsChanged do image.setColor(ch.coords, ch.after)

  def undo(): Unit =
    for ch <- pixelsChanged do image.setColor(ch.coords, ch.before)
}

object ImageChange {
  private case class PixelChange(coords: TriangleCoords, before: Color, after: Color)

  class Builder {
    private val pixelsChanged = ArrayBuffer.empty[PixelChange]

    def done(image: ImageStorage): ImageChange =
      val change = new ImageChange(image, pixelsChanged.reverse.toVector)
      pixelsChanged.clear()
      change

    def addChange(index: TriangleCoords, oldColor: Color, newColor: Color): Builder =
      pixelsChanged += PixelChange(index, oldColor, newColor)
      this

    def nonEmpty: Boolean = pixelsChanged.nonEmpty
  }

  def builder(): Builder = new Builder
}

File: core/src/main/kotlin/tripaint/grid/ImageChange.kt
package tripaint.grid

import tripaint.Change
import tripaint.color.Color
import tripaint.coords.TriangleCoords
import tripaint.image.ImageStorage

class ImageChange(private val image: ImageStorage, private val pixelsChanged: List<PixelChange>) : Change {
    override fun redo() {
        for (ch in pixelsChanged) {
            image.setColor(ch.coords, ch.after)
        }
    }

    override fun undo() {
        for (ch in pixelsChanged) {
            image.setColor(ch.coords, ch.before)
        }
    }

    data class PixelChange(val coords: TriangleCoords, val before: Color, val after: Color)

    class Builder {
        private val pixelsChanged = mutableListOf<PixelChange>()

        fun done(image: ImageStorage): ImageChange {
            val change = ImageChange (image, pixelsChanged.reversed())
            pixelsChanged.clear()
            return change
        }

        fun addChange(index: TriangleCoords, oldColor: Color, newColor: Color): Builder {
            pixelsChanged += PixelChange(index, oldColor, newColor)
            return this
        }

        fun nonEmpty(): Boolean = !pixelsChanged.isEmpty()
    }

    companion object {
        fun builder(): Builder = Builder()
    }
}
